### PR TITLE
New Replcation and transaction Modules

### PR DESCRIPTION
**Pull request by prakashngit**
This is a replacement for PR 205 (closed now).  This PR implements two features:

1.  Facility to replicate change-set (new blocks in the state after update) to the storage services associated with all provisioned enclaves for the contract
2. Facility to submit transactions asynchronously.

Both replication and transaction submission are performed after obtaining a response for the update operation being carried out. The two operations are together referred to as a commit operation. The client can simply invoke the commit operation (after obtaining the response to update operation) to carry out both replication and transaction submission asynchronously.  The client can any time in the future to check the status of the commit operation. The commit operation permits the client to specify past commits as dependencies, in case the transaction corresponding to the commit operation must only be submitted after the submission of transactions corresponding to commit operations  specified as dependencies.

Changes with respect to PR 205:

1. Replication thread model has changed: In the new PR, there are a fixed number of threads that handle replication to any given storage service. This helps prevent thread explosion that might have occurred in PR 205.
2. Separate module for transaction: In PR 205, transaction submission functionality was part of the response module. This has now been separated for improved clarity.
3. Changes to address comments on PR205: There were several valuable comments on PR 205, these have been addressed suitably in this PR.

Points to note:

The asynchronous commit is  illustrated via test-request and test-contract. The pdo-update and pdo-shell scripts, while use the new commit modules, carry out "overall commit operation" synchronously. It is expected that pod-sell will be modified in a separate PR to take advantage of asynchronous commits.

Signed-off-by: prakashngit <prakash.narayana.moorthy@intel.com>

### CODE REVIEW

File: client/pdo/client/controller/commands/send.py [hunk 1]
```diff
@@ -81,17 +81,26 @@ def send_to_contract(state, save_file, message, eservice_url=None, quiet=False,
...
+        contract.set_state(update_response.raw_state)
+
+        # asynchronously submit the commit task: (a commit task replicates change-set and submits the corresponding transaction)
+        try:
+            update_response.commit_asynchronously(ledger_config, wait=30, use_ledger=True)
+        except Exception as e:
+            raise Exception('failed to submit commit: %s', str(e))
+
+        # wait for the commit to finish.
+        # TDB: 1. make wait_for_commit a separate shell command. 2. Add a provision to specify commit dependencies as input to send command.
```
Comment (g2flyer) [inline client/pdo/client/controller/commands/send.py]: TDB? TBD?? (If so TODO seems more common tag?)

File: client/pdo/client/controller/commands/create.py [hunk 2]
```diff
@@ -90,9 +90,19 @@ def __create_contract(ledger_config, client_keys, preferred_eservice_client, ese
 
     logger.info('Contract state created successfully')
 
-    logger.info('Saving the initial contract state in the ledger...')
-
-    cclinit_result = initialize_response.submit_initialize_transaction(ledger_config, wait=30.0)
+    # submit the commit task: (a commit task replicates change-set and submits the corresponding transaction)
+    try:
+        initialize_response.commit_asynchronously(ledger_config, wait=30, use_ledger=True)
```
Comment (g2flyer) [inline client/pdo/client/controller/commands/create.py]: Is this 'wait' the timeout after which the async worker gives up?  If so, maybe timeout might be a more meaningful name?  Or is this a wait for getting into the queue to the worker?

If the former, how would a time-out manifest laterin wait_for_commit?

File: client/pdo/client/scripts/UpdateCLI.py [hunk 3]
```diff
@@ -292,18 +313,18 @@ def Main() :
...
     if options.eservice_db:
-        config['Service']['EnclaveServiceDatabaseFile'] = options.eservice_db    
+        config['Service']['EnclaveServiceDatabaseFile'] = options.eservice_db
     if options.enclave :
         if options.enclave == 'random' :
             options.enclave = random.choice(config['Service'].get('EnclaveServiceURLs',['http://localhost:7001']))
         config['Service']['PreferredEnclaveService'] = options.enclave
         # we will not use database
         config['Service']['EnclaveServiceNames'] = []
-
```
Comment (cmickeyb) [inline client/pdo/client/scripts/UpdateCLI.py]: thank you by the way!

File: python/pdo/contract/__init__.py [hunk 4]
```diff
@@ -20,7 +20,15 @@
...
-    "add_enclave_to_contract"
+    "add_enclave_to_contract",
+    "ReplicationRequest",
+    "start_replication_service",
+    "stop_replication_service",
+    "add_replication_task",
+    "TransactionRequest",
+    "start_transaction_processing_service",
+    "stop_transacion_processing_service",
+    "add_transaction_task"
```
Comment (cmickeyb) [inline python/pdo/contract/__init__.py]: normally these would be exported by the various modules (and the __all__ just refers to the modules)

File: python/pdo/contract/contract.py [hunk 5]
```diff
@@ -97,23 +99,35 @@ def __init__(self, code, state, contract_id, creator_id, **kwargs) :
...
+    # -------------------------------------------------------
+    def set_replication_parameters(self, num_provable_replicas=1, availability_duration=120):
+
+        replication_params = dict()
+        replication_params['max_num_replicas'] = len(self.enclave_map.keys())
+        replication_params['num_provable_replicas'] = num_provable_replicas
+        replication_params['availability_duration'] = availability_duration #seconds
+        replication_params['service_ids'] = self.enclave_map.keys() #we replicate to storage services associated with all provisioned encalves
+
+        self.replication_params = replication_params
```
Comment (cmickeyb) [inline python/pdo/contract/contract.py]: why not just write the previous values directly into self.replication_params?

File: python/pdo/contract/state.py [hunk 6]
```diff
@@ -186,6 +188,7 @@ def __pull_blocks_from_eservice__(cls, eservice, contract_id, block_ids, data_di
         block_count = len(blocks_to_pull)
         if block_count == 0 :
             logger.debug('no blocks to pull')
+            logger.info("Pulled 0 new blocks after contract update")
```
Comment (cmickeyb) [inline python/pdo/contract/state.py]: no reason to have two consecutive logger statements

File: python/pdo/service_client/__init__.py [hunk 7]
```diff
@@ -17,5 +17,4 @@
     'generic',
     'provisioning',
     'storage',
-    'servicedatabase'
 ]
```
Comment (cmickeyb) [inline python/pdo/service_client/__init__.py]: why did you remove this?

File: python/pdo/test/contract.py [hunk 8]
```diff
@@ -182,82 +245,32 @@ def CreateAndRegisterContract(config, enclave, contract_creator_keys) :
...
-    except Exception as e :
-        logger.error('failed to add state encryption key; %s', str(e))
-        ErrorShutdown()
+    #add replication information to contract
+    AddReplicationParamsToContract(config, enclaves, contract)
 
-    contract.set_state_encryption_key(enclave.enclave_id, encrypted_state_encryption_key)
+    # Decide if the contract use a fixed enclave or a randomized one for each update. If fixed, we chose here. If random, nothing to be done here,
+    # will be (atuomatically) selected at random during request creation
+    if (use_eservice is False) or (config['Service']['Randomize_Eservice'] is False):
```
Comment (cmickeyb) [inline python/pdo/test/contract.py]: this is a rather complicated way to get at the conditional. could invert... if use_eservice and config['Service']['Randomize_EService'] :

File: client/pdo/client/controller/commands/send.py (hunk 1 above)
Comment (prakashngit) [reply client/pdo/client/controller/commands/send.py]: i always thought "To Be Done" as a common tag. interesting to know that it is not so common!

File: python/pdo/test/contract.py [hunk 9]
```diff
@@ -298,12 +307,20 @@ def CreateAndRegisterContract(config, enclave, contract_creator_keys) :
...
+def UpdateTheContract(config, enclaves, contract, contract_invoker_keys) :
+    commit_dependenices = []
+    last_response_committed = None
 
     ledger_config = config.get('Sawtooth')
     contract_invoker_id = contract_invoker_keys.identity
 
+    # Decide if the contract use a fixed enclave or a randomized one for each update. If fixed, we chose here. If random, nothing to be done here,
+    # will be (atuomatically) selected at random during request creation
+    if (use_eservice is False) or (config['Service']['Randomize_Eservice'] is False):
```
Comment (cmickeyb) [inline python/pdo/test/contract.py]: see above...

File: python/pdo/test/contract.py [hunk 10]
```diff
@@ -483,6 +510,8 @@ def Main() :
     parser.add_argument('--eservice-db', help='json file for eservice database', type=str)
     parser.add_argument('--eservice-name', help='List of enclave services to use. Give names as in database', nargs='+')
     parser.add_argument('--eservice-url', help='List of enclave service URLs to use', nargs='+')
+    parser.add_argument('--randomize-eservice', help="Say True if eservice must be randomized for each update. \
+        Else, the same eservice (the first one in the list of input eservices) will be used for all udpates.", type=bool, default=False)
```
Comment (cmickeyb) [inline python/pdo/test/contract.py]: this is not the way to get a boolean parameter. should be
action=store_true

File: client/pdo/client/controller/commands/send.py (hunk 1 above)
Comment (cmickeyb) [reply client/pdo/client/controller/commands/send.py]: create an issue rather than leave it in the code

File: client/pdo/client/controller/commands/create.py [hunk 11]
```diff
@@ -90,9 +90,19 @@ def __create_contract(ledger_config, client_keys, preferred_eservice_client, ese
...
+    except Exception as e:
```
[the remark this answers is not in the document]
Comment (prakashngit) [reply client/pdo/client/controller/commands/create.py]: Two points: wait here does not make sense perhaps, as it gives an indication that commit will wait. Second: i agree wait (the sawtooth wait) can be 0, just that we have to change the implementation of wait_for_transaction. We can take this discussion offline. Agree with the need for documentation.

File: client/pdo/client/controller/commands/send.py [hunk 12]
```diff
@@ -81,17 +81,26 @@ def send_to_contract(state, save_file, message, eservice_url=None, quiet=False,
...
+        # 3. Return commit_id after send command back to shell so as to use as input commit_dependency in a future send command
```
[the remark this answers is not in the document]
Comment (prakashngit) [reply client/pdo/client/controller/commands/send.py]: Sure, have created issue 216

File: python/pdo/contract/contract.py (hunk 5 above)
Comment (prakashngit) [reply python/pdo/contract/contract.py]: one field is new, but I guess you are right, i can first copy and just add the new field , will do

File: client/pdo/client/controller/commands/create.py (hunk 2 above)
Comment (prakashngit) [reply client/pdo/client/controller/commands/create.py]: the wait is the timeout after which the async worker (for submitting transactions) simply moves on to the next transaction task. this wait as part of the orignal transaction submitter. 

Thanks for raising this. I realize that wait does not perhaps make sense after the new commit module. it gives a wrong impression that the commit waits for 30 seconds, which is clearly not the case. will fix  this.

File: client/pdo/client/controller/commands/send.py (hunk 1 above)
Comment (prakashngit) [reply client/pdo/client/controller/commands/send.py]: yes , issue 216

File: build/__tools__/run-tests.sh [hunk 13]
```diff
@@ -109,6 +109,7 @@ cd ${SRCDIR}/build
 # -----------------------------------------------------------------
 yell start tests without provisioning or enclave services
 # -----------------------------------------------------------------
+
```
[the remark this answers is not in the document]
Comment (prakashngit) [reply build/__tools__/run-tests.sh]: ok, this line was intentionally added, saw an empty after other yells. so added

File: build/opt/pdo/etc/template/pcontract.toml [hunk 14]
```diff
@@ -52,6 +52,13 @@ ProvisioningServiceURLs = [
     "http://127.0.0.1:7003"
 ]
 
+#--------------------------------------------------
+#Replication -- to maintain state availability
+#-------------------------------------------------
+[Replication]
+NumProvableReplicas=2
+Duration=120 #seconds
+
```
[the remark this answers is not in the document]
Comment (prakashngit) [reply build/opt/pdo/etc/template/pcontract.toml]: agreed on the need for documentation. the whole of commit needs to be documented, created new issue 217

File: python/pdo/service_client/__init__.py (hunk 7 above)
Comment (prakashngit) [reply python/pdo/service_client/__init__.py]: this should not have been ever there.

(we originally had the database part of pdo/service_client, so this was added, but then created a new folder service_data (which has an init for the database). so this ought have been removed.

File: client/pdo/client/controller/commands/create.py (hunk 2 above)
Comment (cmickeyb) [reply client/pdo/client/controller/commands/create.py]: could you add another issue to resolve "wait" in an async replication/commit?

Comment (prakashngit) [reply client/pdo/client/controller/commands/create.py]: Sure, thanks, created issue 218.

File: python/pdo/test/contract.py (hunk 8 above)
Comment (prakashngit) [reply python/pdo/test/contract.py]: sure, thank you, will fix this

File: python/pdo/test/contract.py (hunk 9 above)
Comment (prakashngit) [reply python/pdo/test/contract.py]: sure, thank you, will fix this

File: build/__tools__/run-tests.sh [hunk 15]
```diff
@@ -334,6 +335,25 @@ if [ $? != 0 ]; then
...
+# -----------------------------------------------------------------
+cd ${SRCDIR}/build
+yell run tests for state replication
+say start mock-contract test with replication 3 eservices 2 replicas needed before txn.
+
+try pdo-test-request --ledger ${PDO_LEDGER_URL} \
+    --pservice http://localhost:7001/ http://localhost:7002 http://localhost:7003 \
+    --eservice-url http://localhost:7101/ http://localhost:7102/ http://localhost:7103/ \
+    --logfile __screen__ --loglevel warn --iterations 10 \
+    --num-provable-replicas 2 --availability-duration 100 --randomize-eservice True
```
[the remark this answers is not in the document]
Comment (prakashngit) [reply build/__tools__/run-tests.sh]: thank you, will fix this

File: build/__tools__/run-tests.sh [hunk 16]
```diff
@@ -334,6 +335,25 @@ if [ $? != 0 ]; then
...
+
+say start memory test test with replication 3 eservices 2 replicas needed before txn
+try pdo-test-contract --ledger ${PDO_LEDGER_URL} --contract memory-test \
+    --pservice http://localhost:7001/ http://localhost:7002 http://localhost:7003 \
+    --eservice-url http://localhost:7101/ http://localhost:7102/ http://localhost:7103/ \
+    --logfile __screen__ --loglevel warn \
+    --num-provable-replicas 2 --availability-duration 100 --randomize-eservice True
```
[the remark this answers is not in the document]
Comment (prakashngit) [reply build/__tools__/run-tests.sh]: thank you, will fix this

File: python/pdo/test/contract.py (hunk 10 above)
Comment (prakashngit) [reply python/pdo/test/contract.py]: thank you, will fix this

File: python/pdo/contract/state.py (hunk 6 above)
Comment (prakashngit) [reply python/pdo/contract/state.py]: thank you, removed the info log

File: client/pdo/client/controller/commands/create.py [hunk 17]
```diff
@@ -90,9 +90,19 @@ def __create_contract(ledger_config, client_keys, preferred_eservice_client, ese
...
+        raise Exception('failed to submit commit: %s', str(e))
```
[the remark this answers is not in the document]
Comment (prakashngit) [reply client/pdo/client/controller/commands/create.py]: the exception is handled by the caller __create_contract. (which reraises it). I guess usage here is consistent with other exceptions in this module

File: client/pdo/client/controller/commands/create.py [hunk 18]
```diff
@@ -90,9 +90,19 @@ def __create_contract(ledger_config, client_keys, preferred_eservice_client, ese
...
+
+    # wait for the commit to finish
+    try:
+        txn_id = initialize_response.wait_for_commit()
+        if txn_id is None:
+            raise Exception("Did not receive txn id for the initial commit")
+    except Exception as e:
+        raise Exception("Error while waiting for commit: %s", str(e))
```
[the remark this answers is not in the document]
Comment (prakashngit) [reply client/pdo/client/controller/commands/create.py]: thank you, same reply as above

File: python/pdo/contract/__init__.py (hunk 4 above)
Comment (prakashngit) [reply python/pdo/contract/__init__.py]: ideally i would have liked to export the whole module itself, somehow this did not work, so ended up exporting a whole bunch of methods

File: python/pdo/contract/replication.py [hunk 19]
```diff
@@ -0,0 +1,277 @@
...
+
+# -----------------------------------------------------------------
+__replication_manager_executor__ = concurrent.futures.ThreadPoolExecutor(max_workers=1)
+__pending_replication_tasks_manager_queue__ = queue.Queue()
+
+__replication_workers_executor__ = dict() #key is service id, value is a ThreadPoolExecutor object that manages the worker threads for this storage service
+num_threads_per_storage_service = 2 # we many want to parametrize this later
+__pending_replication_tasks_workers_queues__ = dict() #key is service id, value is a queue of pending replication tasks for this storage service
+
+__condition_variable_for_completed_tasks__ = threading.Condition() # used to notify the parent thread about a new task that got completed (if the parent is waiting)
```
Comment (bvavala) [inline python/pdo/contract/replication.py]: I think this `__condition_variable_for_completed_tasks__ ` works, though it should be in the replication request object.

Long answer.
As I understand it, there may be multiple tasks being processed by the manager and then by the workers.
However, once a replication request completes or fails there is one global condition variable which is updated.

Now, if you have 2 request from 2 threads, which are waiting for them to complete, when the first terminates you notify both of them -- though one will make progress, the other keeps waiting.

File: python/pdo/contract/replication.py [hunk 20]
```diff
@@ -0,0 +1,277 @@
...
+            # check if the overall task can be marked as successful or failed:
+            if len(replication_request.successful_services) >= replication_request.num_provable_replicas:
+                replication_request.mark_as_completed(response.call_back_after_replication)
+            elif len(replication_request.unsuccessful_services) > len(replication_request.service_ids) - replication_request.num_provable_replicas:
+                replication_request.mark_as_failed()
+
+            # Finally, if the task failed, mark the service as unreliable
+            # (this may be a bit harsh, we will refine this later based on the nature of the failure)
+            if fail_task:
+                __services_to_ignore__.add(service_id)
```
[the remark this answers is not in the document]
Comment (prakashngit) [reply python/pdo/contract/replication.py]: thank you for the questions. Answer to question 1. (is set-add thread safe) : My understanding is that it is, My reference is "https://docs.python.org/3/faq/library.html?highlight=atomic" Please see "What kinds of global value mutation are thread-safe?" It specifically says "In practice, it means that operations on shared variables of built-in data types (ints, lists, dicts, etc) that “look atomic” really are". And this I believe includes set.add().
If set.add() is not atomic, yes I agree we need locks!

2. __service_to_ignore__ and unseccessful_service may appear to have some overlap as of now, but they have very different meaning. . unsuccessful_services are those that failed for the current task (due to any reason like e.g. it is a byzantine service !) while __services_to_ignore is a continually updated list of services that we ignore for future tasks (for instance a storage service may experience a temporary link outage or may be genuinely busy for the next 10 seconds (say) and we ingore it for sometime, and add it back later (of course in the current implementation we do not add it back if we ingore, but the point is both are very different in their usage)

File: python/pdo/contract/replication.py [hunk 21]
```diff
@@ -0,0 +1,277 @@
...
+        self.mark_as_completed()
+
+    # -----------------------------------------------------------------
+    def wait_for_completion(self):
+        """ Returns after successful completion of the replication request. Raises exception if the request failed"""
+
+        release_lock = False
+        while self.is_completed is False:
+            __condition_variable_for_completed_tasks__.acquire()
+            __condition_variable_for_completed_tasks__.wait(timeout=1.0)
```
[the remark this answers is not in the document]
Comment (prakashngit) [reply python/pdo/contract/replication.py]: thank you for the question. I agree, a timeout is not necessary here, the code should work the same with or without the timeout.  I will remove the timeout if it clarifies the logic

File: python/pdo/contract/replication.py [hunk 22]
```diff
@@ -0,0 +1,277 @@
...
+            release_lock = True
```
[the remark this answers is not in the document]
Comment (prakashngit) [reply python/pdo/contract/replication.py]: Thank you. I guess this question is related to your previous question. I will remove the timeout as noted above to clarify the logic. (Please note that the lock is released at the start of the wait and reacquired after the wait irrespective of whether you get notified or time out, so I believe the current code works just fine)

File: python/pdo/contract/replication.py [hunk 23]
```diff
@@ -0,0 +1,277 @@
...
+
+            # add the task to the workers queues
+            for service_id in ids_of_services_to_use:
+                __pending_replication_tasks_workers_queues__[service_id].put(response)
+
+    except Exception as e:
+        logger.info("Replication Manager exception %s", str(e))
+
+# -----------------------------------------------------------------
+def __replication_worker__(service_id, pending_taks_queue, condition_variable_for_setup):
```
[the remark this answers is not in the document]
Comment (prakashngit) [reply python/pdo/contract/replication.py]: thank you, will fix the spelling, and the copies of it!

File: python/pdo/contract/replication.py [hunk 24]
```diff
@@ -0,0 +1,277 @@
...
+        self.mark_as_completed()
+
+    # -----------------------------------------------------------------
+    def wait_for_completion(self):
+        """ Returns after successful completion of the replication request. Raises exception if the request failed."""
+
+        release_lock = False
+        while self.is_completed is False:
+            __condition_variable_for_completed_tasks__.acquire()
+            __condition_variable_for_completed_tasks__.wait()
```
Comment (bvavala) [inline python/pdo/contract/replication.py]: Still a problem here:
if you enter the loop and, meanwhile, the other thread sets the variable to true and releases the lock... you will get stuck in the wait.
Solution: put the `is_completed` variable in the critical section, by acquiring the lock before the check.

File: python/pdo/contract/replication.py [hunk 25]
```diff
@@ -0,0 +1,277 @@
...
+            release_lock = True
```
Comment (bvavala) [inline python/pdo/contract/replication.py]: further simplification:
now that there is no timeout, the code will always check for completion, and acquire and release the lock so... you don't need the `release_lock` variable and check anymore

File: python/pdo/contract/replication.py (hunk 24 above)
Comment (prakashngit) [reply python/pdo/contract/replication.py]: thank you! just fixed this

File: python/pdo/contract/replication.py (hunk 25 above)
Comment (prakashngit) [reply python/pdo/contract/replication.py]: thank you, simplified the code in the latest pull.
